### DISCUSSION TITLE
Can I highlight or annotate selected content using a screen reader?

### DISCUSSION THREAD

**Discussion by cary-rowen** [edrlab/thorium-reader#2463] [Q&A] [answered]
I would like to mark certain sentences as highlighted during reading.
And allows me to list all highlighted content and jump to a specific reading position in the chapter.

I also like to add annotations to selected content, which allows me to look back on my reading and find flashes of inspiration.
Sorry, I don't know how to do this with a screen reader.
Are there any guides?

**Comment by danielweck** [comment, marked as answer]
Hello, we are about to publish Thorium v3.0 with a new highlight/annotations feature which takes as input the HTML DOM selection created by the user, as well as some optional user-entered text, colour and style of highlight (solid, outline, strikethrough, underline).
We are aware that some screen readers do not expose native web browsers selection (HTML DOM Range) so in version 3.1 later this year we will implement a workaround that will consist in accepting a text snippet from the user (copied from virtual buffer clipboard) and trying to match the given text with a real HTML DOM Range in the displayed document.

**Comment by cary-rowen** [comment]
Hi @danielweck 
> So in version 3.1 later this year we will implement a workaround that will consist in accepting a text snippet from the user (copied from virtual buffer clipboard) and trying to match the given text with a real HTML DOM Range in the displayed document.

So happy to hear that, that's great.

But is it reliable enough? What happens if there are multiple entries in the HTML DOM Range that match a given content?
If I understand it wrong then I'm sorry.


BTW, NVDA has implemented an app-native selection mode, might this be helpful?
[nvaccess/nvda/pull/15830](https://github.com/nvaccess/nvda/pull/15830)

**Reply by danielweck** [reply]
yes, NVDA now supports native DOM Range in Gecko/Firefox.
However Thorium is built with the Chromium open source web browser engine.

As for the clipboard workaround ... yes absolutely, this is not as accurate / reliable as a real user text selection, this would only be a workaround.

**Comment by nvdaes** [comment]
Hello Daniel, Cary and all:
We need to press NVDA+enter to click on the last position to move to it using the special underscores link in the frame, though I think it's not enough documented for screen reader users.
Wold be possible to click on two positions to highlight or anotate them?
I tested a Colibrio demo using a web browser, which used a copy to clipboard approach to create annotations, years ago, and I didn't like the experience, as I can remember, though this was ling time ago.

**Reply by danielweck** [reply]
Hello Noelia, text selections for annotations are precise ranges of characters, so if the user chooses a "begin" and an "end" element, I think that in most HTML documents of EPUB publications this will just select a huge range of text, probably completely useless for the annotations feature. Furthermore, there is always the ambiguity of selecting text inside the chosen element, or the text that starts immediately after the element closes, etc.
Good to hear your feedback about the clipboard / fuzzy-matching approach which has its own tradeoffs / inconvenients.

**Comment by nvdaes** [comment]
Hi Daniel, then I suppose that the only workaround would be the clipboard approach, unfortunately, since I don't like the idea of putting text on the clipboard mangling other available text.
But I would like to have this feature.
But at least, please, use a comfortable keystroke by default 😄

**Comment by cary-rowen** [comment]
Hi @danielweck 
Has there been some progress here lately?

**Comment by cary-rowen** [comment]
Hello cc @danielweck 
In [this comment](https://github.com/nvaccess/nvda/pull/15830#issuecomment-2573870622), Chromium engineer pointed out that the Chromium bug related to this has been fixed.
Meanwhile, I opened NVDA's feature request. Can we do something in thoriumReader?

https://github.com/nvaccess/nvda/issues/17591

**Reply by danielweck** [reply]
Ideally creating HTML DOM text selections using a screen reader will be as easy as doing it with a mouse / trackpad or other pointing device. Once a native web browser selection exists in the publication document (aka DOM Range), then Thorium simply offers the user the possibility to create an annotation for the selected range of characters. There is an icon button in the top toolbar, there are shortcuts, and there are several annotation modes (by default a modal popup dialog is displayed to enter additional text, choose annotation style and colours, etc., but this can be bypassed in order to annotation faster).

**Reply by danielweck** [reply]
I should point out that Electron doesn't use the very latest version of the Chromium web browser engine, it will take time to reach Thorium.
